### DISCUSSION TITLE
Count + Group By

### DISCUSSION THREAD

**Discussion by kennethsato** [mikro-orm/mikro-orm#5754] [Q&A]
How should I do a count + groupBy in Mikro-ORM?

I'm trying it like this:
`    const total = await this.ormRepository.count(filter, { groupBy: 'status' });`

I'm refactoring a code I had in sequelize that was something like:
`      const total = await Model.count({group: ['status']})`

Expected result would be:
```
{
  countModel: [
    { status: 'emprocesso', count: 20 },
    { status: 'aprovacao', count: 10 },
    { status: 'alteracao', count: 6 },
    { status: 'entregue', count: 201 },
    { status: 'finalizado', count: 14694 }
  ]
}
```

I know count method only returns a 'Promise < number >", so should I call count for each status?

**Comment by B4nan** [comment]
`em.count()` returns a number, you need to use a query builder for this. Something like (not tested):

```ts
const res = await this.em.createQueryBuilder(Model).select(['status', sql`count(id)`).groupBy('status');
```

**Reply by kennethsato** [reply]
Thank you for the answer @B4nan .
I tried something close to your suggestion:
`    const res = await this.ormRepository.createQueryBuilder('Chamada').select(['status', `count(id)`]).groupBy('status');`
But it did not work. I got the following error: 

> [ERROR] 09:52:29 ValidationError: You cannot merge entity 'Chamada' without identifier!

Besides, I don't know where should that sql before count come from...

Probably I'm doing something wrong here, so I tried going to knex:
`     const res = await this.ormRepository.createQueryBuilder('Chamada').select(['status', `count(id)`]).groupBy('status'); `
And the output was:

> [
>   { status: 'finalizado', 'count(`id`)': 16032 },
>   { status: 'emprocesso', 'count(`id`)': 549 },
>   { status: 'aprovacao', 'count(`id`)': 179 },
>   { status: 'alteracao', 'count(`id`)': 208 },
>   { status: 'aprovado', 'count(`id`)': 9 },
>   { status: 'entregue', 'count(`id`)': 395 }
> ]

Which is really close to what I expected.

Could you help me adjusting QueryBuilder to create that?

**Reply by B4nan** [reply]
You need to use qb.execute(), the sql helper is exported from the orm as usual.

**Reply by kennethsato** [reply]
I tried to import from @mikro-orm/core and @mikro-orm/mysql and none of them has sql module.

The execute should be ran after createQueryBuilder, like this?
`const qb = await this.ormRepository.createQueryBuilder('Chamada').select(['status', sql`count(id)`]).groupBy('status');

qb.execute();`

I tried it and qb has no execute because it is a "Chamada" array.
And I still get that validation error when building my query

**Reply by B4nan** [reply]
They both have it, sounds like you are on v5 still if you cant find it.

**Reply by kennethsato** [reply]
Yes. I'm still on v5.
Any chance accomplishing that on v5?
I know I should, but I could not upgrade node or ts yet.

**Reply by B4nan** [reply]
You dont need the `sql` helper in v5, something like this should work:

```ts
const res = await this.ormRepository.createQueryBuilder('Chamada').select(['status', 'count(id)']).groupBy('status').execute();
```

Note the `qb.execute()` - you are not supposed to be calling that on the awaited result, but on the QB instance itself.

**Reply by kennethsato** [reply]
It worked, @B4nan !
Thank you!

I wasn't really expecting an answer for that. Thanks so much for still helping even with older versions.
I'm gonna increase upgrade priority.
